File: src/test/scala/com/avast/kluzo/KluzoTest.scala
package com.avast.kluzo

import java.util.concurrent.{Executor, Executors}

import com.avast.continuity.{Continuity, ContinuityContextThreadNamer}
import com.avast.kluzo.Kluzo._
import org.scalatest.FunSuite
import org.slf4j.LoggerFactory

import scala.concurrent.ExecutionContext

class KluzoTest extends FunSuite {

  /** See the log output to check that traceId is present. */
  private val logger = LoggerFactory.getLogger(this.getClass)

  test("traceId") {
    testTraceId(Executors.newSingleThreadExecutor)
    testTraceId(ExecutionContext.global)
  }

  private def testTraceId(pool: Executor) {
    implicit val threadNamer = ContinuityContextThreadNamer.prefix(ContinuityKey)
    val wrappedPool = Continuity.wrapExecutor(pool)

    val traceId1 = TraceId.generate
    Kluzo.withTraceId(traceId1) {
      wrappedPool.execute(new Runnable {
        override def run(): Unit = {
          assert(getTraceId === Some(traceId1))
          logger.info("first")
        }
      })
    }

    val traceId2 = TraceId.generate
    Kluzo.withTraceId(traceId2) {
      pool.execute(new Runnable {
        override def run(): Unit = {
          assert(getTraceId === None)
          logger.info("second")
        }
      })
    }

    val traceId3 = TraceId.generate
    Kluzo.withTraceId(traceId3) {
      wrappedPool.execute(new Runnable {
        override def run(): Unit = {
          assert(getTraceId === Some(traceId3))
          logger.info("third")
        }
      })
    }
  }
}

File: kluzo/src/test/java/com/avast/kluzo/javaapi/KluzoTest.java
package com.avast.kluzo.javaapi;

import com.avast.continuity.ContinuityContextThreadNamer;
import com.avast.continuity.javaapi.Continuity;
import org.junit.Test;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Optional;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

import static com.avast.kluzo.javaapi.Kluzo.CONTINUITY_KEY;
import static org.junit.Assert.assertEquals;

public class KluzoTest {

    private Logger logger = LoggerFactory.getLogger(this.getClass());

    @Test
    public void testTraceId() throws InterruptedException {
        ExecutorService executor = Executors.newCachedThreadPool();
        ExecutorService wrappedExecutor = Continuity.wrapExecutorService(executor, ContinuityContextThreadNamer.prefix(CONTINUITY_KEY));

        CountDownLatch latch = new CountDownLatch(3);

        TraceId traceId1 = TraceId.generate();
        Kluzo.withTraceId(Optional.of(traceId1), () -> {
            wrappedExecutor.execute(() -> {
                assertEquals(Kluzo.getTraceId(), Optional.of(traceId1));
                logger.info("first");
            });
            latch.countDown();
            return null;
        });

        TraceId traceId2 = TraceId.generate();
        Kluzo.withTraceId(Optional.of(traceId2), () -> {
            executor.execute(() -> {
                assertEquals(Kluzo.getTraceId(), Optional.empty());
                logger.info("second");
            });
            latch.countDown();
            return null;
        });

        TraceId traceId3 = TraceId.generate();
        Kluzo.withTraceId(Optional.of(traceId3), () -> {
            wrappedExecutor.execute(() -> {
                assertEquals(Kluzo.getTraceId(), Optional.of(traceId3));
                logger.info("third");
            });
            latch.countDown();
            return null;
        });

        latch.await();
        Thread.sleep(500);
    }

}
